daemon/api_sideload_n_try.go

```diff
@@ -625,37 +638,98 @@ func readComponentInfoFromContImpl(tempPath string, csi *snap.ComponentSideInfo)
 	return snap.ReadComponentInfoFromContainer(compf, nil, csi)
 }
 
-// readComponentInfo reads ComponentInfo from a snap component file and the
-// snap.Info of the matching installed snap. If instanceName is not empty, it
-// is used to find the right instance, otherwise the SnapName from the
-// component is used.
-func readComponentInfo(st *state.State, tempPath, instanceName string, flags sideloadFlags) (*snap.ComponentInfo, *snap.Info, *apiError) {
-	if !flags.dangerousOK {
-		// TODO:COMPS: read assertions for components
-		return nil, nil, BadRequest("only unasserted installation of local component with --dangerous is supported at the moment")
+func readComponentInfo(st *state.State, upload *uploadedContainer, flags sideloadFlags, model *asserts.Model) (*snap.ComponentInfo, *snap.Info, *apiError) {
+	if flags.dangerousOK {
+		compInfo, err := readComponentInfoFromCont(upload.tmpPath, nil)
+		if err != nil {
+			return nil, nil, BadRequest("cannot read component metadata: %v", err)
+		}
+
+		// unfortunately we must do this a bit out of order, since we cannot
+		// know the component name until opening the container
+		compInfo.ComponentSideInfo = snap.ComponentSideInfo{
+			Component: compInfo.Component,
+			Revision:  snap.R(0),
+		}
+
+		// if no instance was provided in the request we use the snap name from
+		// the component
+		instanceName := upload.instanceName
+		if instanceName == "" {
+			instanceName = compInfo.Component.SnapName
+		}
+
+		info, err := installedSnapInfo(st, instanceName)
+		if err != nil {
+			if errors.Is(err, state.ErrNoState) {
+				return nil, nil, SnapNotInstalled(instanceName, fmt.Errorf("snap owning %q not installed", compInfo.Component))
+			}
+			return nil, nil, BadRequest("cannot retrieve information for %q: %v", instanceName, err)
+		}
+
+		return compInfo, info, nil
```

Comment:
It could be good to separate in two functions, one for the dangerous case and another one for the non-dangerous case.